File: patterns/08_strategy/csharp/ArsBellica.cs
// ============================================================
//  STRATEGY PATTERN — Roman Battle Formations
//  Run: dotnet script ArsBellica.cs
// ============================================================
using System;
record Battle(int Enemies, string Terrain, int Morale);
interface IBattleStrategy { string StrategyName { get; } string Execute(Battle b); }
record Testudo()     : IBattleStrategy { public string StrategyName => "TESTUDO"; public string Execute(Battle b) => $"🐢 Shields locked! vs {b.Enemies} on {b.Terrain}. -60%!"; }
record Cuneus()      : IBattleStrategy { public string StrategyName => "CUNEUS";  public string Execute(Battle b) => $"🔺 CHARGE! Wedge pierces {b.Enemies} on {b.Terrain}!"; }
record Orbis()       : IBattleStrategy { public string StrategyName => "ORBIS";   public string Execute(Battle b) => $"⭕ 360° defence! Morale {b.Morale}/10."; }
record FugaTactica() : IBattleStrategy { public string StrategyName => "FUGA";    public string Execute(Battle b) => $"🏃 Retreat from {b.Enemies} enemies!"; }

class RomanLegion {
    private IBattleStrategy _s;
    private string _name;
    public RomanLegion(string n, IBattleStrategy s) { _name = n; _s = s; }
    public void SetStrategy(IBattleStrategy s) { Console.WriteLine($"  ⚙  {_name}: {_s.StrategyName} → {s.StrategyName}"); _s = s; }
    public void Engage(Battle b) { Console.WriteLine($"\n  🦅 {_name} [{_s.StrategyName}]\n     {_s.Execute(b)}"); }
}

Console.WriteLine("╔═══════════════════════════════════════════════╗");
Console.WriteLine("║   STRATEGY PATTERN — C#                       ║");
Console.WriteLine("╚═══════════════════════════════════════════════╝\n");
var legion = new RomanLegion("Legio X Gemina", new Testudo());
legion.Engage(new Battle(500, "open field", 8));
Console.WriteLine("\n── SCIPIO ADAPTS ────────────────────────────────");
legion.SetStrategy(new Cuneus());      legion.Engage(new Battle(200, "open field", 9));
legion.SetStrategy(new Orbis());       legion.Engage(new Battle(5000, "valley", 4));
legion.SetStrategy(new FugaTactica()); legion.Engage(new Battle(10000, "ambush", 2));
Console.WriteLine("\n\"Non una via vincimus!\"");

File: patterns/08_strategy/typescript/ars_bellica.ts
// ============================================================
//  STRATEGY PATTERN — Roman Battle Formations
//  Run: npx ts-node ars_bellica.ts
// ============================================================

interface Battle { enemies: number; terrain: string; morale: number; }

interface BattleStrategy {
  strategyName(): string;
  execute(b: Battle): string;
}

class Testudo     implements BattleStrategy {
  strategyName() { return "TESTUDO (Tortoise)"; }
  execute(b: Battle) { return `🐢 Shields locked! vs ${b.enemies} on ${b.terrain}. Casualties -60%!`; }
}
class Cuneus      implements BattleStrategy {
  strategyName() { return "CUNEUS (Wedge)"; }
  execute(b: Battle) { return `🔺 CHARGE! Wedge pierces ${b.enemies} on ${b.terrain}!`; }
}
class Orbis       implements BattleStrategy {
  strategyName() { return "ORBIS (Circle)"; }
  execute(b: Battle) { return `⭕ 360° defence! Morale ${b.morale}/10. Holding!`; }
}
class FugaTactica implements BattleStrategy {
  strategyName() { return "FUGA TACTICA"; }
  execute(b: Battle) { return `🏃 Tactical retreat from ${b.enemies} enemies!`; }
}

class RomanLegion {
  constructor(private name: string, private strategy: BattleStrategy) {}
  setStrategy(s: BattleStrategy) {
    console.log(`  ⚙  ${this.name}: ${this.strategy.strategyName()} → ${s.strategyName()}`);
    this.strategy = s;
  }
  engage(b: Battle) {
    console.log(`\n  🦅 ${this.name} [${this.strategy.strategyName()}]`);
    console.log(`     ${this.strategy.execute(b)}`);
  }
}

console.log("╔═══════════════════════════════════════════════╗");
console.log("║   STRATEGY PATTERN — TypeScript               ║");
console.log("╚═══════════════════════════════════════════════╝\n");

const legion = new RomanLegion("Legio X Gemina", new Testudo());
legion.engage({ enemies: 500,   terrain: "open field",         morale: 8 });
console.log("\n── SCIPIO ADAPTS ────────────────────────────────");
legion.setStrategy(new Cuneus());
legion.engage({ enemies: 200,   terrain: "open field",         morale: 9 });
legion.setStrategy(new Orbis());
legion.engage({ enemies: 5000,  terrain: "surrounded valley",  morale: 4 });
legion.setStrategy(new FugaTactica());
legion.engage({ enemies: 10000, terrain: "ambush",             morale: 2 });
console.log('\n"Non una via vincimus!"');
